core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -94,7 +95,13 @@ private[spark] class DiskBlockManager(
       } else {
         val newDir = new File(localDirs(dirId), "%02x".format(subDirId))
         if (!newDir.exists()) {
-          Files.createDirectory(newDir.toPath)
+          // SPARK-37618: Create dir as group writable so files within can be deleted by the
+          // shuffle service
+          val path = newDir.toPath
+          Files.createDirectory(path)
+          val currentPerms = Files.getPosixFilePermissions(path)
+          currentPerms.add(PosixFilePermission.GROUP_WRITE)
```

Comment:
Add `GROUP_READ` and `GROUP_EXECUTE` (for listing) as well.

Reply:
The idea was it should already have everything else it needs (because it needs to to function), so change as little as possible and just add group write. I could just update to fully set as 770 instead

Reply:
Instead of Files.createDirectory, using `DiskBlockManager.createDirWithPermission770` (or move it to utils) should suffice - and we wont need to change the perms.

It is unfortunate that jdk still has not fixed this [bug](https://bugs.openjdk.java.net/browse/JDK-8137404).

Reply:
From what I understand, `createDirWithPermission770` was made because it was thought it couldn't be done through the Java API, but this works in my environment through the Java API correctly. See later in that same thread for that discussion: https://github.com/apache/spark/pull/35085#discussion_r789858724

Because it might be possible to remove that helper after all (still waiting to hear if there's some other reason create then set permission won't work in other environments), I just stuck with the nio API.

From what I've found, losing the `setgid` bit isn't actually Java related. `mkdir -p -m770` also removes the `setgid` bit because of the way the `chmod` Linux system call works (in the operating systems I tested at least). Despite the man page saying that it preservers the setgid bit on directories, the same behavior that the man page lists for files seems to apply to directories. If the user isn't in the group of the directory, it will remove the setgid bit.

Reply:
Now I can't recreate mkdir losing the setgid bit part...

Reply:
This is what I tried:
{code}

mkdir t1 ; chgrp mridul t1; chmod g+s t1; mkdir -m770 t1/t2; mkdir -m770 t1/t2/t3; cp ~/.bashrc t1/t2/t3/bashrc; ls -al t1/t2/
total 12
drwxrws--- 3 test mridul 4096 Jan 29 13:17 .
drwxrwsr-x 3 test mridul 4096 Jan 29 13:17 ..
drwxrws--- 2 test mridul 4096 Jan 29 13:17 t3

{code}

Looks like mkdir honors it.
The issue is that java is overwriting the permission with the new values from setPosixPermissions - and that is blind to the earlier gid bit.

Reply:
Yeah I'll have to get back to you Monday, I'm 99% sure in our production environment even mkdir was losing it, but I'll check again. Thought I could recreate locally in docker containers but can't

Reply:
Thanks @Kimahriman - you are correct, we should get this right. Would be great if we can find out if there are cases where it does not work (which will also mean the existing merge directory creation might have issues perhaps).

Reply:
Ok I figured it out. When you use `mkdir -m<perm>`, one of two things happens:
- If `<perm>` meets the current `umask`, then the `mkdir` system call can create the directory with the permission you tell it
- If `<perm>` doesn't meet the current `umask`, then a `mkdir` system call will create the directory and follow it up with a `chmod` system call to set the permissions to what you tell it

Looks like the default umask for a normal linux user is `0002`, so that's why your example worked and it worked when I tried locally as well. If you set the umask to `0022`, that's when you'll see the inner directories lose the `setgid` bit. And the umask for secure yarn processes is 0027.

The chmod call is what doesn't really follow the man page behavior correctly. It states:
```
Setuid and Setgid Bits
chmod clears the set-group-ID bit of a regular file if the file's group ID does not match the user's effective group ID or one of the user's supplementary group IDs, unless the user has appropriate privileges. Additional restrictions may cause the set-user-ID and set-group-ID bits of MODE or RFILE to be ignored. This behavior depends on the policy and functionality of the underlying chmod system call. When in doubt, check the underlying system behavior.
chmod preserves a directory's set-user-ID and set-group-ID bits unless you explicitly specify otherwise. You can set or clear the bits with symbolic modes like u+s and g-s, and you can set (but not clear) the bits with a numeric mode.
```

It seems to be the first part of the description for "regular files" is what's happening here. If the user you are running `chmod` isn't in the group of the folder you're trying to `chmod`, then modifying the group write bit ends up clearing the setgid bit. If you `strace` the system calls, you see it try to `fchmod 02770`, but this ends up removing the setgid bit in this case for some reason (based on the man page it sounds like a Linux bug to me)

Reply:
Given we are using this for directories, and not files, the required functionality should work ?
Did you observe failure when running it ?

On my linux box it still honors if I specify umask of 0027.

```
$ rm -rf t1; umask 0027; mkdir t1 ; chgrp mridul t1; chmod g+s t1; mkdir -m770 t1/t2; mkdir -m770 t1/t2/t3; cp ~/.bashrc t1/t2/t3/bashrc
$ find t1 | xargs ls -al
-rw-r----- 1 test mridul 3771 Jan 31 12:33 t1/t2/t3/bashrc

t1:
total 12
drwxr-s--- 3 test mridul 4096 Jan 31 12:33 .
drwxr-xr-x 3 test test   4096 Jan 31 12:33 ..
drwxrws--- 3 test mridul 4096 Jan 31 12:33 t2

t1/t2:
total 12
drwxrws--- 3 test mridul 4096 Jan 31 12:33 .
drwxr-s--- 3 test mridul 4096 Jan 31 12:33 ..
drwxrws--- 2 test mridul 4096 Jan 31 12:33 t3

t1/t2/t3:
total 12
drwxrws--- 2 test mridul 4096 Jan 31 12:33 .
drwxrws--- 3 test mridul 4096 Jan 31 12:33 ..
-rw-r----- 1 test mridul 3771 Jan 31 12:33 bashrc
```

Reply:
Take your `test` user out of the `mridul` group and try

Reply:
And the problem is the files in the folder
- The DiskBlockManager creates the blockmgr subdirs using `mkdir -p -m770 blockmgr/00`
- This creates the subdirs as `bob:hadoop` with mode `rwxrwx---` (no setgid bit anymore)
- Spark executor creates a shuffle file `blockmgr/00/shuffle_0_0_0.data` as `bob:bob` with mode `rw-r-----` because the parent dir didn't have the setgid bit anymore
- The shuffle service no longer has permission to read this file so it fails

Reply:
Thanks for clarifying @Kimahriman, that makes sense.
+CC @tgravescs for thoughts.

Reply:
All the same problems exist with the Hadoop file systems, but it can be used instead of java.nio.files. Not really much different.

Reply:
I started to look to see how Tez does this. tez supports removing shuffle files.  I took a quick look and I only saw them using the hadoop classes. I haven't had time to investigate fully where everything is being set or what it is set to but I'm pretty sure they keep the setgid on all the directories for shuffle output.

 I'm not sure exactly which problem you are referring to here but hadoop does deal with setting permissions:
https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/RawLocalFileSystem.java#L971

I don't necessarily care we use there classes but we can look to see how its doing permissions, I want to make sure we aren't opening up more then we need to.

Reply:
The problem I'm talking about is losing the setgid bit when creating a folder with 770 permissions. This problem is the same whether you use mkdir, Files.setPosixPermissions, or FileSystem.setPermissions. A lot of this is probably somewhat specific to running a secure yarn setup:
- Node managers run as `yarn:hadoop` and create all container directories as `<user>:hadoop` with the setgid bit and umask `0027`
- Applications run as the running user as a local Linux user
- These users are _not_ in the `hadoop` group (the superuser group)

Running the spark standalone probably doesn't need all the permissions changes assuming everything is just running as the same user? I don't know where tez fits into this (does it run processes as a local Linux user?)

Reply:
tez runs on yarn just like spark does (hive on tez).  it has an external shuffle service and it supports removing those files once their DAG is complete. Basically does the same thing you are trying to add here.  I took a brief look at their code and I didn't see them doing anything special with permissions like you had to add here and it appeared they are just using the Hadoop filesystem calls so I'm assuming somewhere there is setting the permissions properly and it works just fine in a secure yarn setup

Reply:
I'd have to actually run something to see exactly what it's doing with the shuffle files, but I've tried every which way I can think of and this is the only way I've gotten it working (other than using Linux ACLs)

Reply:
@tgravescs Does Tez not rely on Yarn to remove the application local dirs once the application finishes? If the TEZ DAG corresponds to a Yarn application and if the shuffle data is written under the app local dirs, then probably it is just leveraging Yarn to remove the entire app local dirs. NodeManager deletes the app local dirs as the user using [this code](https://github.com/apache/hadoop/blob/7c97c0f969e6e04513e0b5920855e3b7095e8d2b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/native/container-executor/impl/container-executor.c#L2767), I think. Even for Spark, this data gets deleted when the Spark application finishes but in this change we are trying to remove the data earlier by the Spark shuffle service.

Reply:
A  DAG does not mean the entire application, so it removes the intermediate data before the yarn application finishes.  It is for the same purpose as this PR,  when you have very large or long running jobs it wants to cleanup that data. https://issues.apache.org/jira/browse/TEZ-3362

Reply:
I see the code path for that and I don't see how that can work in a secure yarn environment without the node managers running as root based on what I've seen

Reply:
at the time that patch was added by some of my co-workers and we were running in secure yarn environment so I assume that feature works there.  I'll try to take a look, just been swamped lately.

Reply:
I was looking more at the TEZ change and I'm not seeing how it works either unless they are changing the default umask or I'm just missing something (I'm trying to ask the writers of that).  You either need to modify the permissions or have it do the deletion as the user (which is how the directories are normally cleaned up).  From a security point of view opening up permissions is not ideal.

Reply:
Yeah definitely not ideal, but the permissions of the parent folders should prevent this from actually opening up any security holes I believe

Reply:
YARN has a deletion service that it uses to remove the files normally (as the user), I started to look at this but I'm not sure the shuffle manager can get a handle to it.

If we are going to take this approach, I want to only change permissions when this feature is enabled.  Can we limit it to just when creating the shuffle files?
Ideally I would also like it documented.

Reply:
Yeah I hoped there was a way to hook into the yarn run as user in the shuffle service but I couldn't see anyway to do that

Reply:
I put all the permissions changes inside checks for the feature being enabled. I also removed the permission change from the temp file creation, I _think_ the actual shuffle files only go through the `createTempFileWith`